### PR TITLE
Feat/csm

### PR DESCRIPTION
**Pull request by eddort**

### CODE REVIEW

File: src/staking-router-modules/contracts/lido-locator/lido-locator.service.ts [hunk 1]
```diff
@@ -13,8 +13,14 @@ export class LidoLocatorService {
   ) {}
 
   async getStakingRouter(blockTag: BlockTag) {
-    // if set address via attach this.contract.address will still have the same address
-    this.loggerService.log('Contract locator address', this.contract.address);
+    const locatorENVAddress = this.config.get('LIDO_LOCATOR_ADDRESS');
+
+    this.loggerService.log('Contract locator address', { contract: this.contract.address, env: locatorENVAddress });
```
Comment (Amuhar) [inline src/staking-router-modules/contracts/lido-locator/lido-locator.service.ts:18]: Let's make the names of the contract and environment more descriptive to understand why we have different addresses. Both of them are contracts.

File: src/staking-router-modules/community-module.service.ts [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+  ): Promise<boolean> {
+    return await this.keyRegistryService.operatorsWereChanged(moduleAddress, fromBlockNumber, toBlockNumber);
+  }
+
+  public async updateOperators(moduleAddress: string, blockHash: string): Promise<void> {
+    await this.keyRegistryService.updateOperators(moduleAddress, blockHash);
+  }
+
+  public async getCurrentNonce(moduleAddress: string, blockHash: string): Promise<number> {
+    const nonce = await this.keyRegistryService.getStakingModuleNonce(moduleAddress, blockHash);
```
Comment (Amuhar) [inline src/staking-router-modules/community-module.service.ts:41]: return await this.keyRegistryService.getStakingModuleNonce(moduleAddress, blockHash);

File: src/scripts/extract-csm-abi.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import * as fs from 'fs';
```
Comment (Amuhar) [inline src/scripts/extract-csm-abi.ts:1]: lets use typechain for csm contract too

File: src/common/registry-csm/main/validator-registry/validator-registry.module.ts
Comment (Amuhar) [inline src/common/registry-csm/main/validator-registry/validator-registry.module.ts:1]: what if we remove validator-registry.module validator-registry.service? this module was in library lido-nestjs-modules/registry , in keys api we don't need it

Comment (eddort) [reply src/common/registry-csm/main/validator-registry/validator-registry.module.ts:1]: I think the right thing to do would be to make fully backwards compatible modules

File: src/common/registry-csm/fetch/key-batch.fetch.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    moduleAddress: string,
+    operatorIndex: number,
+    unformattedRecords: KeyBatchRecord,
+    startIndex: number,
+    usedKeysCount: number,
+  ): RegistryKey[] {
+    const keys = this.unformattedKeysToArray(unformattedRecords[0]);
+    const signatures = this.unformattedSignaturesToArray(unformattedRecords[1]);
+
+    if (keys.length !== signatures.length) {
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/key-batch.fetch.ts:43]: done

**Review by gatleas17 — APPROVED** [review]

File: src/common/registry-csm/fetch/key.constants.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const REGISTRY_KEY_BATCH_SIZE = 200;
```
Comment (Amuhar) [inline src/common/registry-csm/fetch/key.constants.ts:1]: looks like we dont use this file

File: src/common/registry-csm/fetch/operator.fetch.ts [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  public async operatorsWereChanged(
+    moduleAddress: string,
+    fromBlockNumber: number,
+    toBlockNumber: number,
+  ): Promise<boolean> {
+    if (fromBlockNumber > toBlockNumber) {
+      return false;
+    }
+
+    const nodeOperatorAddedFilter = this.getContract(moduleAddress).filters['NodeOperatorAdded']();
```
[the remark this answers is not in the document]
Comment (eddort) [reply src/common/registry-csm/fetch/operator.fetch.ts:32]: NodeOperatorRewardAddressSet and NodeOperatorNameSet do not exist in CSM

File: src/common/registry-csm/fetch/key.constants.ts (hunk 5 above)
Comment (eddort) [reply src/common/registry-csm/fetch/key.constants.ts:1]: I'll do the refactoring after testnet1

**Review by Amuhar — APPROVED** [review]

**Pull request merged**